crates/iota-genesis-builder/src/stardust/process_outputs.rs

```diff
@@ -247,97 +198,66 @@ impl HornetFilter<Result<(OutputHeader, Output)>> for ScaleIotaAmountFilter {
-            modified_outputs: vec![],
-            number_of_outcome_outputs: 0,
-            with_metrics,
+                Output::Treasury(_) => (),
+            }
         }
     }
-}
 
-impl HornetFilter<Result<(OutputHeader, Output)>> for UnlockedVestingOutputFilter {
-    fn process(
-        &mut self,
-        output: Result<(OutputHeader, Output)>,
-    ) -> Option<Result<(OutputHeader, Output)>> {
+    fn filter_unlocked_vesting_output(&mut self, output: &Result<(OutputHeader, Output)>) -> bool {
         if let Ok((header, inner)) = output {
             if let Some(address) =
-                get_address_if_vesting_output(&header, &inner, self.snapshot_timestamp_s)
+                get_address_if_vesting_output(header, inner, self.snapshot_timestamp_s)
             {
-                if self.with_metrics {
-                    self.modified_outputs.push(header.output_id());
-                }
+                self.vesting_outputs.push(header.output_id());
                 self.unlocked_address_balances
                     .entry(address)
                     .and_modify(|x| x.balance += inner.amount())
                     .or_insert(OutputHeaderWithBalance {
                         output_header: header.clone(),
                         balance: inner.amount(),
                     });
-                // stop filtering
-                None
-            } else {
-                // continue filtering
-                Some(Ok((header, inner)))
+                return false;
             }
-        } else {
-            // continue filtering
-            Some(output)
         }
+        true
     }
 
-    fn pop_processed(&mut self) -> Option<Result<(OutputHeader, Output)>> {
-        self.unlocked_address_balances
-            .pop_first()
-            .map(|(address, output_header_with_balance)| {
-                if self.with_metrics {
-                    self.number_of_outcome_outputs += 1;
-                }
+    fn filter_voting_output(&mut self, output: &mut Result<(OutputHeader, Output)>) {
+        if let Ok((header, inner)) = output {
+            if is_voting_output(inner) {
+                self.voting_outputs.push(header.output_id());
+                // replace the inner output
+                *inner = BasicOutputBuilder::from(inner.as_basic())
+                    .clear_features()
+                    .finish()
+                    .expect("should be able to create a basic output")
+                    .into()
+            }
+        }
+    }
+}
+
+impl<I> Iterator for HornetFilterIterator<I>
+where
+    I: Iterator<Item = Result<(OutputHeader, Output)>>,
+{
+    type Item = I::Item;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if let Some(output) = self.filter_output() {
+            return Some(output);
+        }
+
+        // Now that we are out of the loop we collect the processed outputs from the
+        // filters
+        let next = self.unlocked_address_balances.pop_first().map(
```

Comment:
In here part of the logic of 1 filter is moved into the `next()` method. Ideally the filter should contain all its logic inside his process/filter method.